### ISSUE TITLE
Optimization causing lots of corefx tests to fail

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#14336]
Repro:
Compile/run this in release against the latest coreclr:
```C#
using System;

class Program
{
    static void Main()
    {
        new ArraySegment<byte>(new byte[512], 512, 0);
    }
}
```
It throws:
```
Unhandled Exception: System.ArgumentException: Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.
   at System.ThrowHelper.ThrowArraySegmentCtorValidationFailedExceptions(Array array, Int32 offset, Int32 count) in c:\Users\stoub\Source\Repos\coreclr\src\mscorlib\src\System\ThrowHelper.cs:line 264
```
but should run successfully.

**Comment by AndyAyersMS** [comment]
Looks like this probably came from #14027. Will update when I know more.

**Comment by stephentoub** [comment]
A recent valid build is producing for this repro:
```
G_M45813_IG01:
       4883EC28             sub      rsp, 40

G_M45813_IG02:
       48B9C0007AEEFE7F0000 mov      rcx, 0x7FFEEE7A00C0
       BA00020000           mov      edx, 512
       E8A846815F           call     CORINFO_HELP_NEWARR_1_VC
       488BC8               mov      rcx, rax
       8B5108               mov      edx, dword ptr [rcx+8]
       81FA00020000         cmp      edx, 512
       720F                 jb       SHORT G_M45813_IG04
       81C200FEFFFF         add      edx, -512
       85D2                 test     edx, edx
       7205                 jb       SHORT G_M45813_IG04

G_M45813_IG03:
       4883C428             add      rsp, 40
       C3                   ret

G_M45813_IG04:
       BA00020000           mov      edx, 512
       4533C0               xor      r8d, r8d
       E8CEACA35E           call     System.ThrowHelper:ThrowArraySegmentCtorValidationFailedExceptions(ref,int,int)
       CC                   int3
```
while the latest is producing:
```
G_M45813_IG01:
       4883EC28             sub      rsp, 40

G_M45813_IG02:
       48B9E8C37AEEFE7F0000 mov      rcx, 0x7FFEEE7AC3E8
       BA00020000           mov      edx, 512
       E8A846835F           call     CORINFO_HELP_NEWARR_1_VC
       488BC8               mov      rcx, rax
       8B5108               mov      edx, dword ptr [rcx+8]
       81FA00020000         cmp      edx, 512
       720D                 jb       SHORT G_M45813_IG04
       81C200FEFFFF         add      edx, -512
       7205                 jb       SHORT G_M45813_IG04

G_M45813_IG03:
       4883C428             add      rsp, 40
       C3                   ret

G_M45813_IG04:
       BA00020000           mov      edx, 512
       4533C0               xor      r8d, r8d
       E8D078A65E           call     System.ThrowHelper:ThrowArraySegmentCtorValidationFailedExceptions(ref,int,int)
       CC                   int3
```
Looks like it's missing a `test edx, edx` instruction.

**Comment by mikedn** [comment]
A fix should be available in #14323. Sorry about that.

**Comment by stephentoub** [comment]
Thanks, @mikedn.

**Comment by jkotas** [comment]
Fixed by #14323
